pyparsing/diagram/__init__.py

```diff
@@ -102,26 +107,32 @@ def resolve_partial(partial: "EditablePartial[T]") -> T:
 
 def to_railroad(
     element: pyparsing.ParserElement,
-    diagram_kwargs: dict = {},
-    vertical: int = None,
+    diagram_kwargs: Optional[Dict] = None,
+    vertical: int = 3,
 ) -> List[NamedDiagram]:
     """
     Convert a pyparsing element tree into a list of diagrams. This is the recommended entrypoint to diagram
     creation if you want to access the Railroad tree before it is converted to HTML
+    :param element:
     :param diagram_kwargs: kwargs to pass to the Diagram() constructor
     :param vertical: (optional)
     """
     # Convert the whole tree underneath the root
-    lookup = ConverterState(diagram_kwargs=diagram_kwargs)
-    _to_diagram_element(element, lookup=lookup, parent=None, vertical=vertical)
+    show_results_names = diagram_kwargs.pop('show_results_names', False) if diagram_kwargs is not None else False
+    lookup = ConverterState(diagram_kwargs=diagram_kwargs or {})
+    _to_diagram_element(element, lookup=lookup, parent=None, vertical=vertical, show_results_names=show_results_names)
 
     root_id = id(element)
     # Convert the root if it hasn't been already
-    if root_id in lookup.first:
-        lookup.first[root_id].mark_for_extraction(root_id, lookup, force=True)
+    if root_id in lookup:
+        lookup[root_id].mark_for_extraction(root_id, lookup, force=True)
 
     # Now that we're finished, we can convert from intermediate structures into Railroad elements
-    resolved = [resolve_partial(partial) for partial in lookup.diagrams.values()]
+    diags = list(lookup.diagrams.values())
+    # collapse out duplicate diags with the same name
+    seen = set()
+    deduped_diags = [d for d in diags if d.name not in seen and (seen.add(d.name) or d)]
```

Comment:
Because `diagrams` is a dictionary indexed by element IDs. Is this not sufficient to enforce uniqueness?

Reply:
id(element) is *too* unique. If an element shows up multiple times with different results names, pyparsing creates a copy of that element. This creates separate elements with separate id's, but the same name and same structure. Deduping by id still shows multiple diagrams for that same name/structure. For the purposes of diagramming, we don't need both copies, so deduping by name clears that up. It *does* put the burden on the parser developer to not duplicate names passed to setName(). And it poses an issue for a parser that might have more than one infixNotation, since that method auto-generates names for the internal recursive elements. Generally not a problem, but there is one name that could be a problem, "base_expression". It's a bug to work on during 3.0.0.c1

Reply:
Oh I see. Well if you're using the name as a unique ID then you might as well change the `diagrams` dictionary to use them as indices. Part of the point of this dictionary is deduplication anyway. Or can we `hash` your elements or some other way that is less likely to collide than using names but still compares by value?

Reply:
I just removed the hard-coded "base_expr" name - the diagram correctly shows the element as part of its containing parent.

Reply:
huh? How does that relate to this deduplication problem?

Reply:
It is a variation on the dedup problem. Since we are deduping on element name, we are relying on the developer not to do something like this:

    integer = Word(nums).setName("integer")
    signed_integer = Regex(r"[+-]\d+").setName("integer")  # should be a different name

Which I don't think is too big a thing to ask. But in infixNotation, my pyparsing code makes the calls to setName(), so it is not under the control of the developer. There are several calls to setName there, composing the name from the terms used in the infixNotation operators - there should be no collisions with these. But when I added a hard-coded "base_expr" name also, then this could create a duplicate if you had 2 or more infixNotations in the same parser, and the user wouldn't be able to fix it. (I might need to come up with a new pyparsing warning if a setName is called with the same name for two different elements.)

I added the "base_expr" when we hadn't fully addressed the Forward containment problem. Once that was resolved, I left it in, bu then realized I didn't need it anymore.

Sorry for being so long-winded. This is another case where there is a lot going on.

Reply:
Okay, I see now. But it still seems to me that this dictionary should be just indexed by element names if that is ultimately what we are using to dedup.

Reply:
Not really a conflict here. We need to keep separate entries by element id in the lookup to keep straight which element has which results name, so they get displayed properly in the parent element's diagram. When we are deduping at the end, we are deduping by the subdiagram _name_ - there is no reason to repeat the same subdiagram just because it occurred in two separate places in two separate elements with the same name and description.